compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -13576,8 +13576,7 @@ OMR::Z::TreeEvaluator::referenceArraycopyEvaluator(TR::Node* node, TR::CodeGener
       }
    else
       {
-      // TODO: There are cases under which the following assert will trigger. These need to be fixed and this assert needs to be re-enabled.
-      // TR_ASSERT(node->getArrayCopyElementType() == TR::Address, "Reference arraycopy element type should be a TR::Address but was '%s'", node->getArrayCopyElementType().toString());
+      TR_ASSERT_FATAL(node->getArrayCopyElementType() == TR::Address, "Reference arraycopy element type should be TR::Address but was '%s'", node->getArrayCopyElementType().toString());
```

Comment:
Yes I have done proper testing on downstream projects with all assertions enabled to make sure we never hit this assert. I have also changed from `TR_ASSERT` to `TR_ASSERT_FATAL` to force a fatal crash if this assert triggers even if we compiled with assertions disabled. The reason for this choice is that missing read barriers under concurrent scavenege is a nightmare to debug as the missed read barrier could result in a stale reference to be propagated through the heap and we may not access this stale reference until much later in the application execution path.

As such debugging these types of issues is very difficult and I'd rather fatally assert if we are ever about to miss a read barrier than spend days debugging such problems in core dumps.